component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -161,58 +160,191 @@ protected void initiateAuthenticationRequest(HttpServletRequest request,
+                                    log.debug("email claim is not available for user " + username);
+                                }
+                                redirectToEmailAddressReqPage(response, context, emailOTPParameters, queryParams,
+                                        username);
+                            } else {
+                                updateEmailAddressForUsername(context, request, username);
+                                email = getEmailValueForUsername(username, context);
+                            }
+                        }
+                        if (StringUtils.isNotEmpty(email)) {
+                            processEmailOTPFlow(request, response, email, username, queryParams, context);
+                        }
                     }
+
+                } else if (sendOtpToFederatedEmail) {
+                    // send email OTP for federated users
+                    if (StringUtils.isEmpty(federatedEmailAttributeKey)) {
+                        federatedEmailAttributeKey = getFederatedEmailAttributeKey(context, stepConfigMap
+                                .get(context.getCurrentStep()).getAuthenticatorList().iterator().next().getName());
+                    }
+
+                    for (Map.Entry<ClaimMapping, String> entry : userAttributes.entrySet()) {
+                        String key = String.valueOf(entry.getKey().getLocalClaim().getClaimUri());
+                        String value = entry.getValue();
+                        if (key.equals(federatedEmailAttributeKey)) {
+                            email = String.valueOf(value);
+                            break;
+                        }
+                    }
+
+                    if (StringUtils.isEmpty(email)) {
+                        if (isEmailOTPMandatory) {
+                            if (log.isDebugEnabled()) {
+                                log.debug("There is no email claim to send OTP");
+                            }
+                            throw new AuthenticationFailedException("There is no email claim to send OTP.");
+                        } else {
+                            if (log.isDebugEnabled()) {
+                                log.debug("Email OTP step is skipped since there is no email claim to send OTP and " +
+                                        "email OTP is not mandatory.");
+                            }
+                            processFirstStepOnly(authenticatedUser, context);
+                        }
+
+                    } else {
+                        context.setProperty(EmailOTPAuthenticatorConstants.RECEIVER_EMAIL, email);
+                        processEmailOTPFlow(request, response, email, username, queryParams, context);
+                    }
+
+                } else if (isEmailOTPMandatory) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Send OTP to federated user is disabled");
```

Comment:
How does isEmailOTPMandatory being true connects to this log? Can you please tell me what I'm missing here?

Reply:
If for a federated user, 'sendOTPToFederatedEmailAttribute' is not enabled then the email will not be sent. But if email OTP is configured as mandatory, then the authentication should be failed. If it is not mandatory, then email OTP can be skipped.